storage/swift_storage.py

```diff
@@ -1,253 +1,254 @@
+        retry_swift_operation(
+            f"Failed to store Swift object {object_name} in container {container}",
+            connection.put_object, container, object_name, in_file, content_type=mimetype)
 
-        for root, _, files in os.walk(source_directory):
-            container_path = root.replace(source_directory, object_name, 1)
+    def load_from_filename(self, in_path: str) -> None:
+        with open(in_path, "rb") as fp:
+            self.load_from_file(fp)
 
-            for file in files:
-                retry.attempt(
-                    self._upload_file,
-                    os.path.join(root, file),
-                    object_path=os.path.join(container_path, file))
+    def delete(self) -> None:
+        connection = self.get_connection()
+        container, object_name = self.get_container_and_object_names()
 
-    def delete(self):
-        self._authenticate()
-        container_name, object_name = self._get_container_and_object_names()
-        self._cloudfiles.delete_object(container_name, object_name)
+        retry_swift_operation(
+            f"Failed to delete Swift object {object_name} in container {container}",
+            connection.delete_object, container, object_name)
 
-    def delete_directory(self):
-        self._authenticate()
-        container_name, object_name = self._get_container_and_object_names()
+    def get_download_url(self, seconds: int = 60, key: Optional[str] = None) -> str:
+        connection = self.get_connection()
 
-        # recursively find and delete all objects below object_name
-        _, objects = self._list_container_objects(container_name, object_name)
+        download_url_key = key or self.download_url_key
 
-        for obj in objects:
-            self._cloudfiles.delete_object(container_name, obj.name)
+        if download_url_key is None:
+            raise SwiftStorageError(
+                "Missing required `download_url_key` for `get_download_url`.")
 
-    def get_download_url(self, seconds=60, key=None):
-        self._authenticate()
-        container_name, object_name = self._get_container_and_object_names()
-        temp_url_key = key if key is not None else self.download_url_key
+        host, _ = connection.get_service_auth()
+        container, object_name = self.get_container_and_object_names()
 
-        download_url = self._cloudfiles.get_temp_url(
-            container_name, object_name, seconds=seconds, method="GET", key=temp_url_key)
+        storage_url, _ = connection.get_auth()
+        storage_path = urlparse(storage_url).path
 
-        parsed_url = urlparse.urlparse(download_url)
-        parsed_url = parsed_url._replace(path=urllib.quote(parsed_url.path))
+        path = swiftclient.utils.generate_temp_url(
+            f"{storage_path}/{container}/{object_name}",
+            seconds=seconds, key=download_url_key, method="GET")
 
-        return urlparse.urlunparse(parsed_url)
+        return cast(str, urljoin(host, path))
```

Comment:
Why `cast`?

Reply:
This is also to make the type checker happy

Reply:
This one shouldn't be necessary for the typechecker, since `urljoin()` is stdlib and typed *however*, it is generic (e.g. AnyStr, etc.) so it will return the type of the first argument I believe (which might be `bytes`, etc.). I think the typechecker was complaining because the `path` result above is calling an untyped method, which means it results in `Any`, and therefore the `urljoin()` result is actually `Any`. If we cast the value of `path` to `str` we should be good to remove this cast.

Reply:
I'm wondering if it is worthwhile providing our own type annotations for just the interfaces we're using, at least until the upstream libraries provide type annotations, or they are added to `typeshed`?

My main concern is really just structural: it feels like including the `cast` calls where they are are mixing concerns. I.e., the `cast`ing has more to do with the library we're calling than the implementation of the `storage` object, and we're mixing those two concerns in one place. Conceptually, the code might be easier to understand if there was a separate place for the hints to the typechecker about what the untyped upstream libraries return.